### PR TITLE
Introduce kotlin coroutines API

### PR DESCRIPTION
**Pull request by shchuko**
Backward compatibility is preserved: blocking API is implemented using `runBlocking { }` wrappers.

`TeamCityInstanceBuilder` is introduced, so I had to deprecate a few methods it replaces. 

hint: I tried to do my best to keep changes iterative and clean, so per-commit review mustn't be hard.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,23 +1,27 @@
...
+    .latest()
 build!!.downloadArtifacts("*.zip", File("out"))
 ```
 
-Another snippet will run a build on your own server
+Another snippet will run a build on your own server using *coroutines non-blocking API*.
 ```kotlin
-val tc = TeamCityInstanceFactory.httpAuth(
-        "https://myserver.local", "login", "password")
+val tc = TeamCityInstanceBuilder("https://myserver.local")
```
Comment (Andronnix) [inline README.md:16]: The advantage to the old example is that it is working against a real server, just need to copy and paste it. Would be great to have another working example with coroutines against teamcity.jetbrains.com

**Review by Andronnix — CHANGES_REQUESTED** [review]
That's a mega job! I'm torn a little bit though if this should be 1.22 or 2.0, as the change is significant

File: teamcity-rest-client-impl/src/main/kotlin/org/jetbrains/teamcity/rest/builder.kt [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    
+    fun buildBlockingInstance(): TeamCityInstance = TeamCityInstanceBlockingBridge(buildCoroutinesInstance())
+
+    internal fun setUrlBaseAndAuthHeader(urlBase: String, authHeader: String?): TeamCityInstanceBuilder {
+        this.urlBase = UrlBase.values().first { it.value == urlBase }
+        this.authHeader = authHeader
+        return this
+    }
+
+    override fun equals(other: Any?): Boolean {
```
[the remark this answers is not in the document]
Comment (shchuko) [reply teamcity-rest-client-impl/src/main/kotlin/org/jetbrains/teamcity/rest/builder.kt:110]: Used to compare builders in DeprecatedFactoryTest

https://github.com/shchuko/teamcity-rest-client/blob/c4e6ffa3bffe94191492b497b487518eb232d109/teamcity-rest-client-impl/src/test/kotlin/org/jetbrains/teamcity/rest/DeprecatedFactoryTest.kt#L25

**Comment by shchuko** [comment]
> I'm torn a little bit though if this should be 1.22 or 2.0, as the change is significant

If we meet [SEMVER](https://semver.org/) rules:
> Given a version number MAJOR.MINOR.PATCH, increment the:
>
> MAJOR version when you make incompatible API changes
> MINOR version when you add functionality in a backward compatible manner
> PATCH version when you make backward compatible bug fixes

There's no need to increment MAJOR version and 1.22 is OK because the changes are backward-compatible.

**Comment by shchuko** [comment]
I've also added retries configuration support, to be actively used at least in IntelliJ build scripts.

**Comment by shchuko** [comment]
Rebased onto recent master

**Pull request merged**